File: model-microservice/rabbitmq/producer/rabbitmqProducer.py
import pika
import json

class RabbitMqProducer:
    def __init__(self, url: str):
        self.url = url
        self.connection = None
        self.channel = None

    def connect(self):
        self.connection = pika.BlockingConnection(pika.URLParameters(self.url))
        self.channel = self.connection.channel()

    def produce(self, queue: str, message: dict):
        if not self.channel:
            raise Exception('Channel is not initialized. Call connect() first.')

        self.channel.queue_declare(queue=queue, durable=True)
        self.channel.basic_publish(
            exchange='',
            routing_key=queue,
            body=json.dumps(message),
            properties=pika.BasicProperties(
                delivery_mode=2, 
            )
        )

    def close(self):
        if self.channel:
            self.channel.close()
        if self.connection:
            self.connection.close()

File: backend-microservice/src/RabbitMq/producer/rabbitmqProducer.ts
import * as amqp from 'amqplib';
import { queueConfig } from '../../config/rabbitmqConfig';

class RabbitMqProducer {
  private connection: amqp.Connection | null = null;
  private channel: amqp.Channel | null = null;
  private url = queueConfig;

  async connect(): Promise<void> {
    this.connection = await amqp.connect(this.url);
    this.channel = await this.connection.createChannel();
  }

  async produce(queue: string, message: any): Promise<void> {
    if (!this.channel) {
      throw new Error('Channel is not initialized. Call connect() first.');
    }

    await this.channel.assertQueue(queue, { durable: true });
    this.channel.sendToQueue(queue, Buffer.from(JSON.stringify(message)), {
      persistent: true,
    });
  }

  async close(): Promise<void> {
    if (this.channel) {
      await this.channel.close();
    }
    if (this.connection) {
      await this.connection.close();
    }
  }
}

export default RabbitMqProducer;
